noir/noir-repo/compiler/noirc_evaluator/src/ssa/ir/instruction.rs

```diff
@@ -599,22 +603,36 @@ impl Instruction {
     }
 }
 
+pub(crate) type ErrorType = HirType;
+
+#[derive(Debug, Copy, PartialEq, Eq, Hash, Clone, PartialOrd, Ord)]
+pub(crate) struct ErrorSelector(u64);
+
+impl ErrorSelector {
+    pub(crate) fn new(typ: &ErrorType) -> Self {
+        match typ {
+            ErrorType::String(_) => Self(STRING_ERROR_SELECTOR),
+            _ => {
+                let mut hasher = FxHasher::default();
+                typ.hash(&mut hasher);
+                let hash = hasher.finish();
+                assert!(hash != 0, "ICE: Error type {} collides with the string error type", typ);
+                Self(hash)
+            }
+        }
+    }
+
+    pub(crate) fn to_u64(self) -> u64 {
+        self.0
+    }
+}
+
 #[derive(Debug, PartialEq, Eq, Hash, Clone)]
 pub(crate) enum ConstrainError {
     // These are errors which have been hardcoded during SSA gen
     Intrinsic(String),
     // These are errors issued by the user
-    UserDefined(UserDefinedConstrainError),
-}
-
-#[derive(Debug, PartialEq, Eq, Hash, Clone)]
-pub(crate) enum UserDefinedConstrainError {
-    // These are errors which come from static strings specified by a Noir program
-    Static(String),
-    // These are errors which come from runtime expressions specified by a Noir program
-    // We store an `Instruction` as we want this Instruction to be atomic in SSA with
-    // a constrain instruction, and leave codegen of this instruction to lower level passes.
-    Dynamic(Instruction),
+    UserDefined(Vec<ValueId>, ErrorSelector),
```

Comment:
This is a nit but I think `ErrorSelector` first makes more sense like in the `AssertionPayload`